File: distributions.nim
import sequtils
import random, random/urandom, random/mersenne


type
  MyRNG = type(initMersenneTwister(urandom(16)))
  # A generic typeclass for a random var
  RandomVar[A] = concept x
    var rng = initMersenneTwister(urandom(16))
    rng.sample(x) is A
  # A few concrete instances
  ConstantVar[A] = object
    value: A
  Uniform = object
    a, b: float
  Discrete[A] = object
    values: ref seq[A]
  ProcVar[A; B; C] = object
    source: ref B
    transform: proc(a: A): C
  ClosureVar[A] = object
    f: proc(rng: var MyRNG): A

# How to sample from various concrete instances
proc sample[A](rng: var RNG, c: ConstantVar[A]): A = c.value

proc sample[A; B; C](rng: var RNG, p: ProcVar[A, B, C]): C =
  p.transform(sample(rng, p.source[]))

proc sample(rng: var RNG, u: Uniform): float = u.a + (u.b - u.a) * rng.random()

proc sample[A](rng: var RNG, d: Discrete[A]): A = rng.randomChoice(d.values[])

proc sample[A](rng: var MyRNG, c: ClosureVar[A]): A = c.f(rng)

# A few constructors
converter constant[A](a: A): ConstantVar[A] = ConstantVar[A](value: a)

proc uniform(a, b: float): Uniform = Uniform(a: a, b: b)

proc choose[A](xs: seq[A]): Discrete[A] =
  new result.values
  result.values[] = xs

# How to lift a function on values to a function on random variables
proc lift1[A; B; C](f: proc(a: A): C, b: B): ProcVar[A, B, C] =
  new result.source
  result.source[] = b
  result.transform = f

template mapper(f: typed, B: typedesc): auto =
  proc inner(rng: var MyRNG): B =
    f(rng.sample(x))

  result.f = inner

proc map[A, B](x: ConstantVar[A], f: proc(a: A): B): ConstantVar[B] =
  result.value = f(x.value)

proc map[A, B](x: Discrete[A], f: proc(a: A): B): ClosureVar[B] =
  mapper(f, B)

proc map[A, B](x: ClosureVar[A], f: proc(a: A): B): ClosureVar[B] =
  mapper(f, B)

# proc lift2[A; B; C; D](f: proc(a: A, b: B): D, c: C): ProcVar[(A, B), C, D] =
#   new result.source
#   result.source[] = c
#   result.transform = f
#
# proc `&`[A, B](x: RandomVar[A], y: RandomVar[B]): ProcVar[(A, B)] =
#   new result.source
#   result.source[] = b
#   result.transform = f

# Other utilities, e.g. the mean:

# Not the most accurate way to compute the mean, but still
proc mean(rng: var RNG, r: RandomVar[float], samples = 10000): float =
  (1 .. samples).foldl(a  + rng.sample(r), 0.0) / samples.float
  # var total = 0.0
  # for _ in 1 .. samples:
  #   total += rng.sample(r)
  # return total / samples.float

# For a more specific types we can have overloads:
proc mean(rng: var RNG, r: Uniform, samples = 10000): float = (r.b - r.a) / 2.0

when isMainModule:
  import typetraits, future

  proc sq(x: float): float = x * x

  # I would like to be able to auto-generate
  # this with a macro
  template sq(x: RandomVar[float]): auto =
    lift1(sq, x)

  # template `*`(x, y: RandomVar[float]): auto =
  #   lift2(`*`, x & y)

  let
    c = constant(3)
    u = uniform(2, 18)
    d = choose(@[1, 2, 3])
    s = sq(u)
    t = d.map((x: int) => x * x)
    z = t.map((x: int) => x * x)

  # I would also like to write
  # (with a different meaning, two different samples)
  # s = u * u
  # This will require a little macro trickery to define overloads
  # for `sq`, `*` and so on

  var rng = initMersenneTwister(urandom(16))

  # Check that they are all random variables
  echo(c is RandomVar[int])
  echo(u is RandomVar[float])
  echo(d is RandomVar[int])
  echo(s is RandomVar[float])
  echo(t is RandomVar[int])
  echo(t is RandomVar)
  # Sampling
  echo rng.sample(c)
  echo rng.sample(s)
  echo rng.sample(t)
  echo rng.sample(z)
  echo rng.mean(s)
  echo rng.mean(u)
  # All this rng is repetitive: another macro would allow
  # to use a common context, like this:
  #
  # withRng rng:
  #   echo sample(c)
  #   echo sample(s)
  #   echo mean(s)
  #
  # Whenever a statement does not compile, try to add rng as first parameter

File: Distributions.scala
package example

import java.util.Random

// A generic typeclass for a random var
trait RandomVar[A] {
  def sample(rng: Random): A
}

object RandomVar {
  // Provides alternate notation:
  // rng.sample(r)
  // becomes
  // r.sample(rng)
  implicit class MyRandom(val rng: Random) {
    def sample[A](r: RandomVar[A]) = r.sample(rng)
  }

  // How to lift a function on values to a function on random variables
  def lift1[A1, B](f: A1 => B) = (r: RandomVar[A1]) =>
    ProcVar(rng => f(r.sample(rng)))

  def lift2[A1, A2, B](f: (A1, A2) => B) = (r1: RandomVar[A1], r2: RandomVar[A2]) =>
    ProcVar(rng => f(r1.sample(rng), r2.sample(rng)))

  // Other utilities, e.g. the mean:
  // Not the most accurate way to compute the mean, but still
  def mean(rng: Random, r: RandomVar[Double], samples: Int = 10000) =
    (1 to samples).foldLeft(0.0)({ (tot, _) => tot + r.sample(rng) }) / samples
}

// A few concrete instances
class Constant[A] private(a: A) extends RandomVar[A] {
  def sample(rng: Random) = a
}

object Constant {
  def apply[A](a: A) = new Constant(a)
}

class Uniform private(a: Double, b: Double) extends RandomVar[Double] {
  def sample(rng: Random) = a + (b - a) * rng.nextDouble()
}

object Uniform {
  def apply(a: Double, b: Double) = new Uniform(a, b)
}

class Discrete[A] private(as: Seq[A]) extends RandomVar[A] {
  def sample(rng: Random) = as(rng.nextInt(as.length))
}

object Discrete {
  def apply[A](as: Seq[A]) = new Discrete(as)
}

class ProcVar[A] private(p: Random => A) extends RandomVar[A] {
  def sample(rng: Random) = p(rng)
}

object ProcVar {
  def apply[A](p: Random => A) = new ProcVar(p)
}


object Distributions extends App {
  import RandomVar._

  def sq(x: Double) = x * x
  def sum(a: Int, b: Int): Int = a + b

  val c = Constant(3)
  val u = Uniform(3, 18)
  val d = Discrete(List(1, 2, 3, 4))
  // This is not ideal yet
  // I would like to write
  // val sq = lift1(sq)
  // but the symbols clash
  val s = lift1(sq)
  // One can overload symbols by using methods, but this
  // still requires too many type annotations
  def sum(r1: RandomVar[Int], r2: RandomVar[Int]): RandomVar[Int] = lift2[Int, Int, Int](sum)(r1, r2)
  val rng = new Random

  println(rng.sample(c))
  println(rng.sample(u))
  println(rng.sample(d))
  println(rng.sample(s(u)))
  println(rng.sample(sum(c, d)))

  println(mean(rng, u))
}